examples/ble-gatt-server/src/main/java/no/nordicsemi/android/ble/ble_gatt_server/GattService.kt

```diff
@@ -0,0 +1,278 @@
+                BleAdvertiser.advertiseData(),
+                bleAdvertiseCallback!!
+        )
+    }
+
+    private fun disableBleServices() {
+        if (bleAdvertiseCallback != null) {
+            val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+            bluetoothManager.adapter.bluetoothLeAdvertiser?.stopAdvertising(bleAdvertiseCallback!!)
+            bleAdvertiseCallback = null
+        }
+
+        serverManager?.close()
+        serverManager = null
+    }
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return DataPlane()
+    }
+
+    /**
+     * Functionality available to clients
+     */
+    inner class DataPlane : Binder() {
+        /**
+         * Change the value of the GATT characteristic that we're publishing
+         */
+        fun setValue(value: String) {
+            serverManager?.setValue(value)
+        }
+    }
+
+    companion object {
+        private const val TAG = "gatt-service"
+    }
+
+    /*
+     * Manages the entire GATT service, declaring the services and characteristics on offer
+     */
+    private class ServerManager(val context: Context) : BleServerManager(context), ServerObserver {
+
+        private val myGattCharacteristic = characteristic(MyServiceProfile.MY_CHARACTERISTIC_UUID,
+                BluetoothGattCharacteristic.PROPERTY_READ // properties
+                        or BluetoothGattCharacteristic.PROPERTY_NOTIFY,
+                BluetoothGattCharacteristic.PERMISSION_READ_ENCRYPTED_MITM,  // permissions
+                description("A characteristic to be read", false) // descriptors
+        )
+        private val myGattService = service(MyServiceProfile.MY_SERVICE_UUID, myGattCharacteristic)
+
+        private val serverConnections = mutableMapOf<String, ServerConnection>()
+
+        private var currentValue: String? = null
+
+        fun setValue(value: String) {
+            currentValue = value
+            serverConnections.values.forEach { client ->
+                client.setValue(value)
+            }
+        }
+
+        override fun log(priority: Int, message: String) {
+            if (BuildConfig.DEBUG || priority == Log.ERROR) {
+                Log.println(priority, TAG, message)
+            }
+        }
+
+        override fun initializeServer(): List<BluetoothGattService> {
+            setServerObserver(this)
+
+            // In this example there's only one service, so singleton list is created.
+            return Collections.singletonList(myGattService)
+        }
+
+        override fun onServerReady() {
+            log(Log.INFO, "Gatt server ready")
+        }
+
+        override fun onDeviceConnectedToServer(device: BluetoothDevice) {
+            log(Log.DEBUG, "Device connected ${device.address}")
+
+            val clientConnection = ServerConnection()
+            clientConnection.useServer(this@ServerManager)
+            clientConnection.connect(device).enqueue()
```

Comment:
Making a connection here allows the BleManagerHandler to be found when a read characteristic is requested by the central. This appears to me as overkill though as the peripheral (this server) shouldn't need to connect and discover the central's services... or do I have this wrong... is it normal that a peripheral connects and discovers the central in return to a central connecting to the peripheral?

Reply:
Separate issue raised: https://github.com/NordicSemiconductor/Android-BLE-Library/issues/237